spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpMessageSource.java

```diff
@@ -0,0 +1,234 @@
+		this.headerMapper = headerMapper;
+	}
+
+	public MessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	public void setMessageConverter(MessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "amqp:message-source";
+	}
+
+	@Override
+	protected AbstractIntegrationMessageBuilder<Object> receiveMessageBuilder() {
+		Connection connection = this.connectionFactory.createConnection();
+		Channel channel = connection.createChannel(this.transacted);
+		try {
+			GetResponse resp = channel.basicGet(this.queue, false);
+			if (resp == null) {
+				RabbitUtils.closeChannel(channel);
+				RabbitUtils.closeConnection(connection);
+				return null;
+			}
+			AcknowledgmentCallback callback = this.ackCallbackFactory
+					.createCallback(() -> new AmqpAckInfo(connection, channel, resp.getEnvelope().getDeliveryTag()));
+			MessageProperties messageProperties = this.propertiesConverter.toMessageProperties(resp.getProps(),
+					resp.getEnvelope(), StandardCharsets.UTF_8.name());
+			Map<String, Object> headers = this.headerMapper.toHeadersFromRequest(messageProperties);
+			Object payload = this.messageConverter
+					.fromMessage(new org.springframework.amqp.core.Message(resp.getBody(), messageProperties));
+			return getMessageBuilderFactory().withPayload(payload)
+					.copyHeaders(headers)
+					.setHeader(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK, callback);
+		}
+		catch (IOException e) {
+			RabbitUtils.closeChannel(channel);
+			RabbitUtils.closeConnection(connection);
+		}
+		return null;
+	}
+
+	public static class AmqpAckCallbackFactory implements AcknowledgmentCallbackFactory<AmqpAckInfo> {
+
+		@Override
+		public AcknowledgmentCallback createCallback(Supplier<AmqpAckInfo> supplier) {
+			return new AmqpAckCallback(supplier.get());
+		}
+
+	}
+
+	public static class AmqpAckCallback implements AcknowledgmentCallback {
+
+		private final AmqpAckInfo ackInfo;
+
+		private volatile boolean acknowledged;
+
+		public AmqpAckCallback(AmqpAckInfo ackInfo) {
+			this.ackInfo = ackInfo;
+		}
+
+		@Override
+		public void acknlowledge(Status status) {
+			Assert.notNull(status, "'status' cannot be null");
+			try {
+				switch (status) {
+				case ACCEPT:
+					this.ackInfo.getChannel().basicAck(this.ackInfo.getDeliveryTag(), false);
+					break;
+				case REJECT:
+					this.ackInfo.getChannel().basicReject(this.ackInfo.getDeliveryTag(), false);
+					break;
+				case REQUEUE:
+					this.ackInfo.getChannel().basicReject(this.ackInfo.getDeliveryTag(), true);
+					break;
+				default:
+					break;
+				}
+			}
+			catch (IOException e) {
+				throw RabbitExceptionTranslator.convertRabbitAccessException(e);
+			}
+			finally {
+				RabbitUtils.closeChannel(this.ackInfo.getChannel());
+				RabbitUtils.closeConnection(this.ackInfo.getConnection());
+			}
+			this.acknowledged = true;
```

Comment:
Is it OK do not mark it as `acknowledged` when we get an exception during this process?

I mean this is outside of the `finally` block, so won't be reachable in case of the exception during `basicAck/basicReject`.

So, any thoughts on the matter, please?

Reply:
Well, strictly speaking if the ack/nack fails then it should remain false. But I agree it's likely to fail again if tried a second time so might just add noise to the log. OK.